### PR TITLE
chore: Makefile improvements to increase developer velocity

### PR DESCRIPTION
**Pull request by ellistarn**
*Issue #, if available:*

*Description of changes:*

* Added a DEVELOPER_GUIDE
* Use ko for image builds, improving build times from 10 minutes -> 23 seconds (on OSX)
* Introduce a `presubmit` target, which executes all steps in the dev workflow
* Introduce a `toolchain` target, which automatically installs all tools required for development (verified in presubmit)
* Includes controller-gen in presubmit, to enforce CRD to be up to date

Local Controller takes lease
```
make apply

k get lease cp-vpc-resource-controller -n kube-system -oyaml
apiVersion: coordination.k8s.io/v1
kind: Lease
metadata:
  creationTimestamp: "2022-12-23T10:23:14Z"
  name: cp-vpc-resource-controller
  namespace: kube-system
  resourceVersion: "62722223"
  uid: 8817e6d3-019a-4068-b363-aea4dc9a2daa
spec:
  acquireTime: "2023-06-29T18:36:56.000000Z"
  holderIdentity: vpc-resource-local-controller-574d58d89c-dmkp4_86162373-4d94-43a7-97b2-275d56dede83
  leaseDurationSeconds: 30
  leaseTransitions: 32
  renewTime: "2023-06-29T18:37:21.585152Z"
```

Controller resumes lease
```
make delete
k get lease cp-vpc-resource-controller -n kube-system -oyaml

apiVersion: coordination.k8s.io/v1
kind: Lease
metadata:
  creationTimestamp: "2022-12-23T10:23:14Z"
  name: cp-vpc-resource-controller
  namespace: kube-system
  resourceVersion: "62721552"
  uid: 8817e6d3-019a-4068-b363-aea4dc9a2daa
spec:
  acquireTime: "2023-06-29T18:33:26.744612Z"
  holderIdentity: ip-10-0-54-3.us-west-2.compute.internal_3159b31f-49d3-42f5-9d6e-36fd56a34566
  leaseDurationSeconds: 30
  leaseTransitions: 32
  renewTime: "2023-06-29T18:35:01.890293Z"```


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: DEVELOPER_GUIDE.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+```sh
+make presubmit
+```
+
+## Troubleshooting
+
+### Invalid value 'trunk' for InterfaceType
+
+The following error means that must be allowlisted for EC2 Networking
```
Comment (jdn5126) [inline DEVELOPER_GUIDE.md:34]: what must be allow-listed?

Comment (haouc) [reply DEVELOPER_GUIDE.md:34]: The trunk API need approval on AWS account level.

File: hack/toolchain.sh [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+KUBEBUILDER_ASSETS="${KUBEBUILDER_ASSETS:="${HOME}/.kubebuilder/bin"}"
+
+main() {
+    tools
+    kubebuilder
+}
+
+tools() {
+    go install sigs.k8s.io/controller-runtime/tools/setup-envtest@v0.0.0-20220421205612-c162794a9b12
+    go install sigs.k8s.io/controller-tools/cmd/controller-gen@v0.6.2
```
[the remark this answers is not in the document]
Comment (ellistarn) [reply hack/toolchain.sh:14]: We can choose @latest, but it means that developers will need to check in new generated yaml (from make presubmit) whenever a new version comes out.

**Review by haouc — APPROVED** [review]
lgtm, thanks for the wonderful improvement. It is much faster for dev.

**Pull request merged**